src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/AbstractCohostDelegatingEndpoint.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.Telemetry;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.Extensions;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
+
+internal abstract class AbstractCohostDelegatingEndpoint<TRequest, TResponse> : AbstractRazorCohostDocumentRequestHandler<TRequest, TResponse?>
```

Comment:
I was avoiding doing any of the AbstractDelegating endpoints because I hadn't thought about how to do it without duplicating this whole class, so needless to say, I don't _love_ this, but at the same time I do recognize that it unblocks a bunch more endpoints. I think at the very least we should have a comment on both classes about mirroring changes, and we should add it to https://github.com/dotnet/razor/issues/9814 to come back to this.

Reply:
Thinking about this more, maybe this is good enough for now, because if we move away from LSP to talk to Roslyn, but instead have services in OOP that use plain C# method calls, most of this goes away for cohosting I guess.